### PR TITLE
84 Create and get points

### PR DESCRIPTION
**Pull request by memeramita**
<h2>Context</h2>

#84 Create functions `create_points_layer` and `get_selected_points` in viewer.py for adding point layers and getting selected points from point layers to support point annotation.

<h2>Changes</h2>

`viewer.py`

- `create_points_layer()`: Create a new Points layer and set it to ADD mode to allow users to add points to the layer.
- `get_selected_points()`: Get selected points in a point layer as a list of point tuples.

I implemented the following class and functions to support `create_points_layer` and `get_selected_points`
- `PointsLayerMode`: This Enum class is added to store Points layer modes.
- `get_all_points_layers()`: A helper function to get all Points layers in the viewer.
- `order_point()`: Rearrange each point's dimension to the same order as the image.
- `set_points_layer_mode()`: Called to set the mode of a Points layer.

`i_viewer.py`

- I added `get_all_points_layers`, `create_points_layer`, and `get_selected_points` to reflect the changes in `viewer.py`

`fake_viewer.py`
- I created `self._points_layers` to keep track of fake Points layers and implemented fake methods `get_all_points_layers`, `create_points_layer`, and `get_selected_points`.

### CODE REVIEW

File: napari_allencell_annotator/_tests/fakes/fake_viewer.py [hunk 1]
```diff
@@ -24,3 +25,14 @@ def alert(self, alert_msg: str) -> None:
...
     def get_layers(self) -> List[Layer]:
         return self._layers
+
+    def get_all_points_layers(self) -> List[Points]:
+        return self._points_layers
+
+    def create_points_layer(self, name: str, color: str, visible: bool) -> Points:
+        points: Points = Points(data=None, name=name, color=color, visible=visible)
+        self._points_layers.append(points)
+        return points
```
Comment (yrkim98) [inline napari_allencell_annotator/_tests/fakes/fake_viewer.py]: this would be closer to the real thing if the points layer was added to the actual layer list. Which is what happens in napari

File: napari_allencell_annotator/_tests/fakes/fake_viewer.py [hunk 2]
```diff
@@ -24,3 +25,14 @@ def alert(self, alert_msg: str) -> None:
...
+
+    def get_selected_points(self, point_layer: Points, image_dims_order: str) -> List[Tuple]:
+        return [(0.0, 0.0, 0.0, 0.0, 0.0)]
```
Comment (yrkim98) [inline napari_allencell_annotator/_tests/fakes/fake_viewer.py]: this should return points.data

File: napari_allencell_annotator/view/viewer.py [hunk 3]
```diff
@@ -44,4 +58,97 @@ def alert(self, alert_msg: str) -> None:
...
+        """
         return list(self.viewer.layers)
+
+    def get_all_points_layers(self) -> List[Points]:
+        """
+        Returns a list of all point layers in the viewer.
+        """
+        return [layer for layer in self.get_layers() if isinstance(layer, Points)]
+
+    @staticmethod
```
Comment (yrkim98) [inline napari_allencell_annotator/view/viewer.py]: avoid staticmethods whenever possible

File: napari_allencell_annotator/view/viewer.py [hunk 4]
```diff
@@ -44,4 +58,97 @@ def alert(self, alert_msg: str) -> None:
...
+            A point in a point layer
+        image_dims_order: str
+            The dimension of the image
+
+        Returns
+        -------
+        Tuple[float]
+            A tuple containing ordered point coordinates
+        """
+        point_dict: Dict[str, np.ndarray] = {"T": point[0], "C": point[1], "Z": point[2], "Y": point[3], "X": point[4]}
```
Comment (yrkim98) [inline napari_allencell_annotator/view/viewer.py]: cant assume order of points, and cant assume that there will be 5 dimensions

Comment (memeramita) [reply napari_allencell_annotator/view/viewer.py]: assume that the order is correct for now and remove the ordering in https://github.com/aics-int/napari-allencell-annotator/pull/90/commits/d8c46768ffb6869970a3be00beae75c0fa68e6de

File: napari_allencell_annotator/view/viewer.py (hunk 3 above)
Comment (memeramita) [reply napari_allencell_annotator/view/viewer.py]: make it a class method in https://github.com/aics-int/napari-allencell-annotator/pull/90/commits/4b48d1d02ffcaed68cc7909fa35f6ae0d7899928

File: napari_allencell_annotator/_tests/fakes/fake_viewer.py (hunk 2 above)
Comment (memeramita) [reply napari_allencell_annotator/_tests/fakes/fake_viewer.py]: fixed in https://github.com/aics-int/napari-allencell-annotator/pull/90/commits/2a9abc6c6a135518aa1673b84d810f0c63fab2d0

File: napari_allencell_annotator/_tests/fakes/fake_viewer.py (hunk 1 above)
Comment (memeramita) [reply napari_allencell_annotator/_tests/fakes/fake_viewer.py]: changed in https://github.com/aics-int/napari-allencell-annotator/pull/90/commits/39abde6e5970471afef7c4250b554f73feda98be

File: napari_allencell_annotator/view/viewer.py [hunk 5]
```diff
@@ -44,4 +73,79 @@ def alert(self, alert_msg: str) -> None:
         show_info(alert_msg)
 
     def get_layers(self) -> List[Layer]:
+        """
+        Returns a list of all layers in the viewer.
+        """
         return list(self.viewer.layers)
+
+    def get_all_points_layers(self) -> List[Points]:
```
Comment (yrkim98) [inline napari_allencell_annotator/view/viewer.py:67]: this looks like a private method

File: napari_allencell_annotator/util/image_utils.py [hunk 6]
```diff
@@ -35,8 +35,8 @@ def __init__(self, filepath: Path):
...
             self._image = BioImage(filepath, reader=bioio_imageio.Reader)
 
-    def get_dask_data(self) -> da.Array:
+    def get_image(self) -> BioImage:
         """
-        Returns image data as a dask array
+        Returns Bioimage object loaded with correct reader
         """
-        return self._image.get_dask_stack()
+        return self._image
```
Comment (yrkim98) [inline napari_allencell_annotator/util/image_utils.py]: Why was this changed back to an image?

File: napari_allencell_annotator/view/viewer.py (hunk 5 above)
Comment (memeramita) [reply napari_allencell_annotator/view/viewer.py:67]: changed to private in https://github.com/aics-int/napari-allencell-annotator/pull/90/commits/ec029a71e9c03314bb78a974d9f09d4299abff5d

File: napari_allencell_annotator/_tests/view/viewer_test.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+@pytest.fixture
+def viewer(make_napari_viewer: napari.Viewer) -> Viewer:
+    return Viewer(make_napari_viewer())
+
+
+def test_all_points_layer(viewer: Viewer) -> None:
+    # ARRANGE
+    viewer.viewer.add_shapes()
+    test_points_layer1: Points = viewer.create_points_layer("test1", "blue", True, 2)
+    test_points_layer2: Points = viewer.create_points_layer("test2", "blue", True, 2)
```
Comment (yrkim98) [inline napari_allencell_annotator/_tests/view/viewer_test.py:25]: We want to avoid mutating fixtures here: https://sapegin.me/blog/avoid-mutation/#:~:text=Mutation%20may%20lead%20to%20unexpected,vigilant%20when%20reading%20the%20code.

File: napari_allencell_annotator/_tests/view/viewer_test.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+
+@pytest.fixture
+def viewer(make_napari_viewer: napari.Viewer) -> Viewer:
+    return Viewer(make_napari_viewer())
+
+
+def test_all_points_layer(viewer: Viewer) -> None:
+    # ARRANGE
+    viewer.viewer.add_shapes()
```
Comment (yrkim98) [inline napari_allencell_annotator/_tests/view/viewer_test.py]: why are we using an `add_shapes()` method here?

we either need to implement this in our viewer, or use a different method to ensure we're only grabbing points layers. I think adding an empty image should suffice.

File: napari_allencell_annotator/_tests/view/viewer_test.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    assert test_points_layer1 in all_points_layer
+    assert test_points_layer2 in all_points_layer
+
+
+def test_add_image(viewer: Viewer) -> None:
+    # ARRANGE
+    test_path: Path = (
+        Path(napari_allencell_annotator.__file__).parent / "_tests" / "assets" / "image_types" / "img.ome.tiff"
+    )
+    test_image: BioImage = ImageUtils(test_path).get_image()
```
Comment (yrkim98) [inline napari_allencell_annotator/_tests/view/viewer_test.py]: No need to load in this image into memory. We can just initialize a small numpy array (2x2 or so) and add that image instead

File: napari_allencell_annotator/_tests/view/viewer_test.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+def test_create_points_layer(viewer: Viewer) -> None:
+    # ACT
+    test_points_layer: Points = viewer.create_points_layer("test", "blue", True, 2)
+
+    # ASSERT
+    assert test_points_layer in viewer.get_all_points_layers()
+    assert len(viewer.get_all_points_layers()) == 1
+    assert test_points_layer.name == "test"
+    assert test_points_layer.visible
+    assert test_points_layer.ndim == 2
```
Comment (yrkim98) [inline napari_allencell_annotator/_tests/view/viewer_test.py:55]: remove dimension from `create_points_layer()`

File: napari_allencell_annotator/_tests/view/viewer_test.py [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+    assert len(test_points_layer.data) == 0
+    assert test_points_layer.mode == PointsLayerMode.ADD.value
+
+
+def test_create_points_layer_color(viewer: Viewer) -> None:
```
Comment (yrkim98) [inline napari_allencell_annotator/_tests/view/viewer_test.py]: move this into the function above

File: napari_allencell_annotator/_tests/view/viewer_test.py [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+
+def test_set_points_layer_mode(viewer: Viewer) -> None:
+    # ARRANGE
+    test_points_layer: Points = viewer.create_points_layer("test", "blue", True, 2)
+
+    # ACT
+    viewer.set_points_layer_mode(test_points_layer, PointsLayerMode.ADD)
+
+    # ASSERT
+    assert test_points_layer.mode == PointsLayerMode.ADD.value
```
Comment (yrkim98) [inline napari_allencell_annotator/_tests/view/viewer_test.py]: use a non-default mode to see if `set_points_layer_mode` works

File: napari_allencell_annotator/_tests/view/viewer_test.py [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+
+
+def test_get_selected_points(viewer: Viewer) -> None:
+    # ARRANGE
+    test_points_layer: Points = viewer.create_points_layer("test", "blue", True, 2)
+    points: np.array = np.array([[1, 2], [3, 4]])
+    test_points_layer.data = points
```
Comment (yrkim98) [inline napari_allencell_annotator/_tests/view/viewer_test.py]: we should not access `data` of the points layer
optional parameter in `create_points_layer` that takes in a list of coordinates and adds it to the data field for testing

File: napari_allencell_annotator/view/i_viewer.py [hunk 14]
```diff
@@ -10,7 +11,7 @@ def __init__(self):
         super().__init__()
 
     @abstractmethod
-    def add_image(self, image: np.ndarray) -> None:
+    def add_image(self, image: BioImage) -> None:
         pass
```
Comment (yrkim98) [inline napari_allencell_annotator/view/i_viewer.py]: change back to dask array

File: napari_allencell_annotator/view/viewer.py [hunk 15]
```diff
@@ -1,30 +1,59 @@
...
+        #         # add each scene separately
+        #         data: dask.array.Array = image.get_image_dask_data(image.dims.order.replace("S", ""), S=i)
+        #         layer = self.viewer.add(data)
+        # else:
+        # # for all other images <=5 dims
+        #     layer = self.viewer.add(image.get_dask_stack())
+
+        self.viewer.add_image(image.get_dask_stack())
+
+        # layer.axis_labels = image.dims.order
```
Comment (yrkim98) [inline napari_allencell_annotator/view/viewer.py]: this should take in a numpy array

File: napari_allencell_annotator/view/i_viewer.py (hunk 14 above)
Comment (yrkim98) [reply napari_allencell_annotator/view/i_viewer.py]: *numpy

File: napari_allencell_annotator/view/viewer.py [hunk 16]
```diff
@@ -44,4 +73,79 @@ def alert(self, alert_msg: str) -> None:
...
+        ndim: int
+            The number of image dimensions
+
+        Returns
+        -------
+        Points
+            A new point layer
+        """
+        points_layer: Points = self.viewer.add_points(None, name=name, face_color=color, visible=visible, ndim=ndim)
+        self.set_points_layer_mode(points_layer=points_layer, mode=PointsLayerMode.ADD)
```
Comment (yrkim98) [inline napari_allencell_annotator/view/viewer.py]: `add` being default points layer doesnt make sense. Dont set mode to begin with

File: napari_allencell_annotator/view/viewer.py (hunk 15 above)
Comment (memeramita) [reply napari_allencell_annotator/view/viewer.py]: fixed in https://github.com/aics-int/napari-allencell-annotator/pull/90/commits/5a1eac3472eb10cd133a20dd0b9a299845829cba

File: napari_allencell_annotator/view/i_viewer.py (hunk 14 above)
Comment (memeramita) [reply napari_allencell_annotator/view/i_viewer.py]: changed back in https://github.com/aics-int/napari-allencell-annotator/pull/90/commits/5a1eac3472eb10cd133a20dd0b9a299845829cba

File: napari_allencell_annotator/util/image_utils.py (hunk 6 above)
Comment (memeramita) [reply napari_allencell_annotator/util/image_utils.py]: Changed back to dask stack in https://github.com/aics-int/napari-allencell-annotator/pull/90/commits/5a1eac3472eb10cd133a20dd0b9a299845829cba

File: napari_allencell_annotator/_tests/view/viewer_test.py (hunk 9 above)
Comment (memeramita) [reply napari_allencell_annotator/_tests/view/viewer_test.py]: changed in https://github.com/aics-int/napari-allencell-annotator/pull/90/commits/34583a5cf38b893ef52818fb136db11ad5df4ff6

File: napari_allencell_annotator/_tests/view/viewer_test.py (hunk 10 above)
Comment (memeramita) [reply napari_allencell_annotator/_tests/view/viewer_test.py:55]: removed in https://github.com/aics-int/napari-allencell-annotator/pull/90/commits/91b886b85e3754b6e4ed509d04bf4334c7819c19

Comment (memeramita) [reply napari_allencell_annotator/_tests/view/viewer_test.py:55]: I removed `ndim` from `create_points_layer()` but I still have to pass ndim=6 when I call `self.viewer.add_points()`. Otherwise, point coordinates would be saved as 2D.

File: napari_allencell_annotator/view/viewer.py (hunk 16 above)
Comment (memeramita) [reply napari_allencell_annotator/view/viewer.py]: removed in https://github.com/aics-int/napari-allencell-annotator/pull/90/commits/2bcd2a279c66ff37011ee8faf6844cff6323f4a8

File: napari_allencell_annotator/_tests/view/viewer_test.py (hunk 12 above)
Comment (memeramita) [reply napari_allencell_annotator/_tests/view/viewer_test.py]: we later decided to no longer set ADD as the default mode

File: napari_allencell_annotator/_tests/view/viewer_test.py (hunk 11 above)
Comment (memeramita) [reply napari_allencell_annotator/_tests/view/viewer_test.py]: moved in https://github.com/aics-int/napari-allencell-annotator/pull/90/commits/461b0287a1ba67e3483968bf34b42bbe29fcc7b7

File: napari_allencell_annotator/_tests/view/viewer_test.py (hunk 13 above)
Comment (memeramita) [reply napari_allencell_annotator/_tests/view/viewer_test.py]: fixed in https://github.com/aics-int/napari-allencell-annotator/pull/90/commits/461b0287a1ba67e3483968bf34b42bbe29fcc7b7

File: napari_allencell_annotator/_tests/view/viewer_test.py (hunk 8 above)
Comment (memeramita) [reply napari_allencell_annotator/_tests/view/viewer_test.py]: https://github.com/aics-int/napari-allencell-annotator/pull/90/commits/39fdc696e96c353c6b3ae8f020ce73eacd2b55a6 I use a (2, 2) image to test since it says ValueError: Image data must have at least 2 dimensions when I use an empty image.

File: napari_allencell_annotator/_tests/view/viewer_test.py (hunk 10 above)
Comment (yrkim98) [reply napari_allencell_annotator/_tests/view/viewer_test.py:55]: discussed- get dims from viewer

File: napari_allencell_annotator/_tests/fakes/fake_viewer.py [hunk 17]
```diff
@@ -24,3 +24,23 @@ def alert(self, alert_msg: str) -> None:
 
     def get_layers(self) -> List[Layer]:
         return self._layers
+
+    def _get_all_points_layers(self) -> List[Points]:
+        return [layer for layer in self.get_layers() if isinstance(layer, Points)]
+
+    def create_points_layer(self, name: str, color: str, visible: bool, data: np.ndarray = None) -> Points:
+        points: Points = Points(data=None, name=name, color=color, visible=visible, ndim=6)
```
Comment (yrkim98) [inline napari_allencell_annotator/_tests/fakes/fake_viewer.py]: `data=data`

File: napari_allencell_annotator/_tests/view/viewer_test.py [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+    assert test_points_layer2 in all_points_layer
+
+
+def test_add_image(viewer: Viewer) -> None:
+    # ACT
+    viewer.add_image(np.zeros(shape=(2, 2)))
+
+    # ASSERT
+    assert len(viewer.get_layers()) == 1
+    np.testing.assert_array_equal(viewer.get_layers()[0].data, np.zeros(shape=(2, 2)))
```
Comment (yrkim98) [inline napari_allencell_annotator/_tests/view/viewer_test.py]: hold reference to image added to assert that the objects are equal later

File: napari_allencell_annotator/_tests/view/viewer_test.py [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+
+
+def test_create_points_layer(viewer: Viewer) -> None:
+    # ACT
+    test_points_layer: Points = viewer.create_points_layer("test", "blue", True, np.zeros(shape=(1, 6)))
+
+    # ASSERT
+    assert test_points_layer in viewer._get_all_points_layers()
+    assert len(viewer._get_all_points_layers()) == 1
```
Comment (yrkim98) [inline napari_allencell_annotator/_tests/view/viewer_test.py]: public function

File: napari_allencell_annotator/_tests/view/viewer_test.py (hunk 10 above)
Comment (memeramita) [reply napari_allencell_annotator/_tests/view/viewer_test.py:55]: fixed in https://github.com/aics-int/napari-allencell-annotator/pull/90/commits/0e863856e340a7c5a741f3669ad9db62993beb65

File: napari_allencell_annotator/_tests/view/viewer_test.py (hunk 19 above)
Comment (memeramita) [reply napari_allencell_annotator/_tests/view/viewer_test.py]: changed in https://github.com/aics-int/napari-allencell-annotator/pull/90/commits/34cae500e0098750b6dec4da800ba0e7f3470a10

File: napari_allencell_annotator/_tests/view/viewer_test.py (hunk 18 above)
Comment (memeramita) [reply napari_allencell_annotator/_tests/view/viewer_test.py]: hold reference in https://github.com/aics-int/napari-allencell-annotator/pull/90/commits/d3daabdc5239f19c6f2959a38f3728e7ab4d6b68

File: napari_allencell_annotator/_tests/fakes/fake_viewer.py (hunk 17 above)
Comment (memeramita) [reply napari_allencell_annotator/_tests/fakes/fake_viewer.py]: fixed in https://github.com/aics-int/napari-allencell-annotator/pull/90/commits/1797d93635cd7c64458a93d0022e3d4cc6f53157

**Review by yrkim98 — APPROVED** [review]
